### PR TITLE
Safer (unicode-safe) flaky test report

### PR DESCRIPTION
**Pull request by potrebic**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @potrebic has signed the CLA. Thanks for the pull request!

File: flaky/utils.py [hunk 1]
```diff
@@ -9,8 +11,20 @@
     unicode_type = str
 
 
-def ensure_unicode_string(string):
+def ensure_unicode_string(obj):
+    """
+    Return a unciode string representation of the given obj.
```
Comment (jmoldow) [inline flaky/utils.py]: Spelling: "unicode"

File: flaky/utils.py [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 from __future__ import unicode_literals
 
+from six import binary_type
```
Comment (jmoldow) [inline flaky/utils.py]: I like `six`, but it doesn't make sense to add a new dependency just for this one use.

Right below this, we're manually determining the `unicode_type`. If the `try` succeeds, then we are in Python 2 and can set `binary_type` to `str`. And if it fails, then we are in Python 3 and can set `binary_type` to `bytes`.

File: flaky/utils.py [hunk 3]
```diff
@@ -9,8 +11,20 @@
...
+        `unicode`
+    """
     try:
-        return unicode_type(string)
+        return unicode_type(obj)
     except UnicodeDecodeError:
-        return string.decode('utf-8', 'replace')
+        if isinstance(obj, binary_type):
+            return obj.decode('utf-8', 'replace')
+        return ensure_unicode_string(repr(obj))
```
Comment (jmoldow) [inline flaky/utils.py]: Before resorting to `repr`, should we try the `binary_type` first?

What about
```python
  except UnicodeDecodeError:
    try:
      return binary_type(obj).decode('utf-8', 'replace')
    except:
      return ensure_unicode_string(repr(obj))
```

File: test/test_example.py [hunk 4]
```diff
@@ -89,13 +89,28 @@ def test_flaky_function(param=[]):
...
+@flaky
 @genty
-class ExampleFlakyTestsWithUnicodeTestNames(ExampleFlakyTests):
-    @genty_dataset('ascii name', 'ńőń ȁŝćȉȉ ŝƭȕƒƒ')
-    def test_non_flaky_thing(self, message):
+class ExampleFlakyTestsWithTrickyTestNames(TestCase):
+    _threshold = -1
+
+    _unicode_string = 'Plain Hello'
+    _byte_string = str(_unicode_string)
```
Comment (jmoldow) [inline test/test_example.py]: Needs to be `binary_type`, not `str`.

If we remove the `six` dependency, maybe move the definitions of `binary_type` and `unicode_type` into a util module, so they can also be accessed here.

File: test/test_example.py [hunk 5]
```diff
@@ -89,13 +89,28 @@ def test_flaky_function(param=[]):
...
+    @genty_dataset(
+        _unicode_string,
+        _byte_string,
+        _unicode_string_non_ascii,
+        _byte_string_non_ascii,
+        _byte_string_windows_encoded,
+    )
+    def test_tricky_strings_in_test_method_name(self, message):
         # pylint:disable=unused-argument
         self._threshold += 1
```
Comment (jmoldow) [inline test/test_example.py]: `_threshold` should be set to -1 in `setUp()`, otherwise all the tests but the first will be no-ops.

**Pull request closed without merging**